service/src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java

```diff
@@ -0,0 +1,130 @@
+            throw new ValidatorException(certificateDTOValidator.getErrorMessage());
+        }
+        String name = certificateDTO.getName();
+        if (certificateRepository.getByName(name).isPresent()) {
+            throw new EntityAlreadyExistsException(" (name = " + name + ")");
+        }
+        Certificate certificate = CertificateConverterDTO.convertToEntity(certificateDTO);
+        checkTags(certificate);
+        certificate = certificateRepository.save(certificate);
+        tagRepository.createCertificateTags(certificate);
+        certificate = certificateRepository.getById(certificate.getId()).orElseThrow(EntityNotFoundException::new);
+        addTagsToCertificate(certificate);
+        return CertificateConverterDTO.convertToDto(certificate);
+    }
+
+    @Override
+    @Transactional
+    public CertificateDTO update(CertificateDTO certificateDTO) {
+        certificateDTOValidator.cleanErrorMessages();
+        if (!certificateDTOValidator.isValid(certificateDTO)) {
+            throw new ValidatorException(certificateDTOValidator.getErrorMessage());
+        }
+        String name = certificateDTO.getName();
+        if (certificateRepository.getByName(name).isPresent()
+                && !name.equals(certificateRepository.getById(certificateDTO.getId()).get().getName())) {
+            throw new EntityAlreadyExistsException(" (name = " + name + ")");
+        }
+        Certificate certificate = CertificateConverterDTO.convertToEntity(certificateDTO);
+        certificateRepository.deleteCertificateTags(certificate.getId());
+        checkTags(certificate);
+        certificate = certificateRepository.update(certificate);
+        tagRepository.createCertificateTags(certificate);
+        certificate = certificateRepository.getById(certificate.getId()).orElseThrow(EntityNotFoundException::new);
+        addTagsToCertificate(certificate);
+        return CertificateConverterDTO.convertToDto(certificate);
+    }
+
+    private void checkTags(Certificate certificate) {
+        Set<Tag> tags = certificate.getTags();
+        if (!tags.isEmpty()) {
+            tagRepository.createNewTags(tags);
+        }
+    }
+
+    @Override
+    public boolean remove(Long id) {
+        return certificateRepository.deleteById(id);
+    }
+
+    @Override
+    public List<CertificateDTO> executeQueryDTO(CertificatePageQueryDTO queryDTO) {
+        pageQueryDTOValidator.cleanErrorMessages();
+        if (!pageQueryDTOValidator.isValid(queryDTO)) {
+            throw new ValidatorException(pageQueryDTOValidator.getErrorMessage());
+        }
+        CertificateParamsRequestUtil paramsRequestUtil = new CertificateParamsRequestUtil(queryDTO);
+        String whereSQL = paramsRequestUtil.getWhereQueryWithParams();
+        String sortSQL = paramsRequestUtil.getSortQueryWithParams();
+        SqlSpecification specification = new CertificateAllSpecification(whereSQL, sortSQL);
```

Comment:
It should look like new `CertificateAllSpecification(queryDTO)` and all the rest will be inside the specification.